### PR TITLE
Back up and restore SecurityDescriptors on Windows

### PR DESCRIPTION
**Pull request by aneesh-n**
What does this PR change? What problem does it solve?
-----------------------------------------------------
Restic did not back up SecurityDescriptors of files on Windows.
Restic now backs up and restores SecurityDescriptors (which includes owner, group, discretionary access control list (DACL), system access control list (SACL)) when backing up files and folders on Windows. This requires the user to be a member of backup operators or the application must be run as admin.
If that is not the case, only the current user's owner, group and DACL will be backed up and during restore only the DACL of the backed file will be restored while the current user's owner and group will be set during the restore.

We may want to add a backup/restore flag for disabling Security descriptor for Windows as there may be valid cases where users do not want to backup/restore security descriptors of files in Windows. 
This can be done as a separate PR.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Follow-up to https://github.com/restic/restic/pull/4611

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/fs/sd_windows.go
Comment (aneesh-n) [inline internal/fs/sd_windows.go:1]: Would have to move this helper into another PR to reduce the total changed lines for this PR.

File: internal/restic/node_windows.go [hunk 1]
```diff
@@ -114,7 +117,7 @@ func (s statT) mtim() syscall.Timespec {
 
 func (s statT) ctim() syscall.Timespec {
 	// Windows does not have the concept of a "change time" in the sense Unix uses it, so we're using the LastWriteTime here.
-	return syscall.NsecToTimespec(s.LastWriteTime.Nanoseconds())
+	return s.mtim()
```
Comment (aneesh-n) [inline internal/restic/node_windows.go:120]: This fixes func statT.mtim is unused (U1000)go-staticcheck and removes code duplication

File: internal/restic/node.go [hunk 2]
```diff
@@ -719,12 +721,7 @@ func (node *Node) fillExtra(path string, fi os.FileInfo) error {
 	allowExtended, err := node.fillGenericAttributes(path, fi, stat)
 	if allowExtended {
 		// Skip processing ExtendedAttributes if allowExtended is false.
-		errEx := node.fillExtendedAttributes(path)
-		if err == nil {
-			err = errEx
-		} else {
-			debug.Log("Error filling extended attributes for %v at %v : %v", node.Name, path, errEx)
-		}
```
Comment (aneesh-n) [inline internal/restic/node.go:727]: Error from fillExtendedAttributes was getting eaten up after this code was updated in https://github.com/restic/restic/pull/4611/files#diff-ed902b2edb54dfe5e342c38a7e7d122cacc6b1e84e956dfc8689b5efd7922fd4

Comment (aneesh-n) [reply internal/restic/node.go:727]: I can move this to the next PR for extended attrs if needed.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this PR, somehow it always ended up in my "it would take too long to review the PR today" category :-( . I've added quite a few comments below, although most of them are just nitpicks.

The only larger issue is the encoding of the SECURITY_DESCRIPTOR struct. I don't think it's a good idea to store the raw data structure. Instead, I think the PR should use the SDDL encoding, see the comment for more details.

File: internal/fs/sd_windows_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	"encoding/base64"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/test"
+)
+
+func Test_SetGetFileSecurityDescriptors(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows_test.go:16]: Done

File: internal/fs/sd_windows.go [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+	var langID uint32
+	err = lookupPrivilegeDisplayName("", &nameBuffer[0], &displayNameBuffer[0], &displayBufSize, &langID)
+	if err != nil {
+		return fmt.Sprintf("<unknown privilege %s>", string(utf16.Decode(nameBuffer[:bufSize])))
+	}
+
+	return string(utf16.Decode(displayNameBuffer[:displayBufSize]))
+}
+
+func adjustTokenPrivileges(token windows.Token, releaseAll bool, input *byte, outputSize uint32, output *byte, requiredSize *uint32) (success bool, err error) {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:390]: Probably, these were copied from winio.

File: internal/fs/sd_windows.go [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+	return privileges, nil
+}
+
+// enableProcessPrivileges enables privileges globally for the process.
+func enableProcessPrivileges(names []string) error {
+	return enableDisableProcessPrivilege(names, SE_PRIVILEGE_ENABLED)
+}
+
+// DisableProcessPrivileges disables privileges globally for the process.
+func DisableProcessPrivileges(names []string) error {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:329]: Had kept these incase we need to disable externally based on flags or something. It was unused, I've removed them now.

File: internal/fs/sd_windows.go [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+	}
+	for i, p := range e.privileges {
+		if i != 0 {
+			s += ", "
+		}
+		s += `"`
+		s += getPrivilegeName(p)
+		s += `"`
+	}
+	if backupPrivilegeError != nil {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:296]: Fixed

File: internal/fs/sd_windows.go [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+	}
+	token := windows.Token(0)
+	member, err := token.IsMember(sid)
+	if err != nil {
+		return false, errors.Errorf("token membership error: %s", err)
+	}
+	return member, nil
+}
+
+// The code below was adapted from github.com/Microsoft/go-winio under MIT license.
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:208]: Done

File: internal/fs/sd_windows.go [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+)
+
+// PrivilegeError represents an error enabling privileges.
+type PrivilegeError struct {
+	privileges []uint64
+}
+
+// SecurityDescriptorBytesToStruct converts the security descriptor bytes representation
+// into a pointer to windows SECURITY_DESCRIPTOR.
+func SecurityDescriptorBytesToStruct(sd []byte) (*windows.SECURITY_DESCRIPTOR, error) {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:267]: Moved them up

File: internal/fs/sd_windows.go [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+	procLookupPrivilegeValueW       = modadvapi32.NewProc("LookupPrivilegeValueW")
+	procAdjustTokenPrivileges       = modadvapi32.NewProc("AdjustTokenPrivileges")
+	procLookupPrivilegeDisplayNameW = modadvapi32.NewProc("LookupPrivilegeDisplayNameW")
+	procLookupPrivilegeNameW        = modadvapi32.NewProc("LookupPrivilegeNameW")
+)
+
+// Do the interface allocations only once for common
+// Errno values.
+const (
+	errnoErrorIOPending = 997
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:243]: This is from windows - https://github.com/microsoft/go-winio/blob/3c9576c9346a1892dee136329e7e15309e82fb4f/tools/mkwinsyscall/mkwinsyscall.go#L948

File: changelog/unreleased/pull-4708 [hunk 10]
```diff
@@ -0,0 +1,12 @@
+Enhancement: Back up and restore SecurityDescriptors on Windows
+
+Restic did not back up SecurityDescriptors of files on Windows.
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply changelog/unreleased/pull-4708:3]: Done

File: doc/040_backup.rst [hunk 11]
```diff
@@ -481,12 +481,17 @@ written, and the next backup needs to write new metadata again. If you really
 want to save the access time for files and directories, you can pass the
 ``--with-atime`` option to the ``backup`` command.
 
+Backing up full security descriptors on windows is only possible when the user 
+has ``SeBackupPrivilege``privilege or is running as admin. This is a restriction 
+of windows not restic.
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply doc/040_backup.rst:519]: Done

File: internal/restic/node_windows.go [hunk 12]
```diff
@@ -23,6 +23,9 @@ type WindowsAttributes struct {
 	CreationTime *syscall.Filetime `generic:"creation_time"`
 	// FileAttributes is used for storing file attributes for windows files.
 	FileAttributes *uint32 `generic:"file_attributes"`
+	// SecurityDescriptor is used for storing security descriptors which includes
+	// owner, group, discretionary access control list (DACL), system access control list (SACL))
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/restic/node_windows.go:27]: Fixed

File: internal/fs/sd_windows.go [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+func enableRestorePrivilege() {
+	err := enableProcessPrivileges([]string{SeRestorePrivilege, SeSecurityPrivilege, SeTakeOwnershipPrivilege})
+	if err != nil {
+		restorePrivilegeError = fmt.Errorf("error enabling restore/security privilege: %w", err)
+	}
+}
+
+// DisableBackupPrivileges disables privileges that are needed for backup operations.
+// They may be reenabled if GetSecurityDescriptor is called again.
+func DisableBackupPrivileges() error {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:168]: I was thinking these would be useful helpers in future, but I've removed them now.

File: internal/fs/sd_windows.go [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+
+// IsAdmin checks if current user is an administrator.
+func IsAdmin() (isAdmin bool, err error) {
+	var sid *windows.SID
+	err = windows.AllocateAndInitializeSid(&windows.SECURITY_NT_AUTHORITY, 2, windows.SECURITY_BUILTIN_DOMAIN_RID, windows.DOMAIN_ALIAS_RID_ADMINS,
+		0, 0, 0, 0, 0, 0, &sid)
+	if err != nil {
+		return false, errors.Errorf("sid error: %s", err)
+	}
+	token := windows.Token(0)
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:200]: windows.Token(0) directly creates a token object representing the access token of the current process with a specific access level. It's a direct way to obtain a token without explicitly opening a handle to the process token.
https://github.com/golang/go/issues/28804

Also, I have tested this.

File: internal/fs/sd_windows.go [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+	// Use a type assertion to check if the error is of type syscall.Errno
+	if errno, ok := err.(syscall.Errno); ok {
+		// Compare the error code to the expected value
+		return errno == windows.ERROR_PRIVILEGE_NOT_HELD
+	}
+	return false
+}
+
+// IsAdmin checks if current user is an administrator.
+func IsAdmin() (isAdmin bool, err error) {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:193]: Done

File: internal/fs/sd_windows.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:267]: In Microsoft's go-winio package (check https://github.com/microsoft/go-winio/blob/bc421d9108ade8ae2f76c8ef3e90cd4ee690e2c0/backuptar/tar.go#L130), they have deprecated sddl-based usage of SecurityDescriptors and instead use base64 encoded raw descriptors. 

I had updated my implementation from sddl to raw accordingly.
"Maintaining old SDDL-based behavior for backward compatibility. All new tar headers written by this library will have raw binary for the security."

Refer - https://github.com/microsoft/go-winio/commit/0ab2de7e65cca04fcdd6bc8846b0d6e6f2732bd7
"Remove use of SDDL in tar headers
Switching to using straight binary instead of SDDL in tar headers to avoid failures in converting domain SIDs into/out of SDDL."

While testing SDs, they were not getting restored for domain accounts when using SSDL and it worked properly only after updating it to use raw binary as specified in go-winio.

File: internal/fs/sd_windows.go [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+	var sd *windows.SECURITY_DESCRIPTOR
+
+	if lowerPrivileges {
+		sd, err = getNamedSecurityInfoLow(sd, err, filePath)
+	} else {
+		sd, err = getNamedSecurityInfoHigh(sd, err, filePath)
+	}
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
+			lowerPrivileges = true
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:56]: If there are concurrent calls, they will all try with the higher privilege and fallback to testing with getNamedSecurityInfoLow and eventually all calls will use either lower or higher privilege.
Is sync really needed for every call for this?

File: internal/fs/sd_windows.go [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+
+	var sd *windows.SECURITY_DESCRIPTOR
+
+	if lowerPrivileges {
+		sd, err = getNamedSecurityInfoLow(sd, err, filePath)
+	} else {
+		sd, err = getNamedSecurityInfoHigh(sd, err, filePath)
+	}
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:55]: Even if it can, I guess there's no point in the fallback for this case. 
I will add - !lowerPrivileges && isHandlePrivilegeNotHeldError(err)

File: internal/restic/node_windows_test.go [hunk 18]
```diff
@@ -12,10 +13,66 @@ import (
...
+	sdInputBytes, err := base64.StdEncoding.DecodeString(sd)
+	test.OK(t, errors.Wrapf(err, "Error decoding SD for: %s", fileName))
+	// Wrap the security descriptor bytes in windows attributes and convert to generic attributes.
+	genericAttributes, err := WindowsAttrsToGenericAttributes(WindowsAttributes{CreationTime: nil, FileAttributes: nil, SecurityDescriptor: &sdInputBytes})
+	test.OK(t, errors.Wrapf(err, "Error constructing windows attributes for: %s", fileName))
+	// Construct a Node with the generic attributes.
+	expectedNode := getNode(fileName, fileType, genericAttributes)
+
+	// Restore the file/dir and restore the meta data including the security descriptors.
+	testPath, node := restoreAndGetNode(t, tempDir, expectedNode, false)
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/restic/node_windows_test.go:43]: Yes, I considered this. I did test that it does not cause conflicts.
This case should be covered under node_windows_test - TestRestoreSecurityDescriptors right?

File: internal/fs/sd_windows.go [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+	// SeTakeOwnershipPrivilege allows the application to take ownership of files and directories, regardless of the permissions set on them.
+	SeTakeOwnershipPrivilege = "SeTakeOwnershipPrivilege"
+
+	backupPrivilegeError  error
+	restorePrivilegeError error
+	lowerPrivileges       bool
+)
+
+// Flags for backup and restore with admin permissions
+var highSecurityFlags windows.SECURITY_INFORMATION = windows.OWNER_SECURITY_INFORMATION | windows.GROUP_SECURITY_INFORMATION | windows.DACL_SECURITY_INFORMATION | windows.SACL_SECURITY_INFORMATION | windows.LABEL_SECURITY_INFORMATION | windows.ATTRIBUTE_SECURITY_INFORMATION | windows.SCOPE_SECURITY_INFORMATION | windows.BACKUP_SECURITY_INFORMATION | windows.PROTECTED_DACL_SECURITY_INFORMATION | windows.PROTECTED_SACL_SECURITY_INFORMATION
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:33]: According to docs it does seem like BACKUP_SECURITY_INFORMATION should cover all, but on removing those other flags like PROTECTED_DACL_SECURITY_INFORMATION, TestSetGetFileSecurityDescriptors fails saying the input and output DACLs do not match. If the flags are added back, the DACLs are restored properly.

The unprotected flags are for acls not protected from being modified by inheritance, which are for auditing purposes.
I will add those flags as well.

[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:33]: Do you think we should have a flag to be able to disable backup of security attributes? 
I think there may be some valid cases when users want to backup files and just have default security attributes used when restored.

File: internal/restic/node_windows_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/restic/node_windows_test.go:43]: Yeah, that test should notice if the chmod would interfere with the security descriptors.

File: internal/fs/sd_windows.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: Thanks for the link.

File: internal/fs/sd_windows.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: using `os.OpenFile` should be fine. By now, Go should also handle most cases on its own that once required `fixpath(...)`. And I guess it's unlikely that users input some crazy path when trying to get debug logs from restic.

File: internal/fs/sd_windows.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: I don't see much reason to not include the security attributes in the backup. But we could discuss about a `restore` option.

File: internal/fs/sd_windows.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go:291]: The function still seems to exist.

File: internal/fs/sd_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go:352]: I've found the source file: https://github.com/microsoft/go-winio/blob/main/zsyscall_windows.go . _I don't want to see copied code ever again that does not clearly state where it originates from._ It's just ridiculous that I have to search for the original source of these functions.

Omitting that information also forces me to waste time on reviewing known good code.

To make matter worse, there have been a few small changes to those functions (apparently to fix linter warnings), which when compared to the latest upstream version of the autogenerated code seem to be buggy. See my other comments for details.

File: internal/fs/sd_windows.go [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+// Do the interface allocations only once for common
+// Errno values.
+const (
+	errnoErrorIOPending = 997
+
+	//revive:disable-next-line:var-naming ALL_CAPS
+	SE_PRIVILEGE_ENABLED = windows.SE_PRIVILEGE_ENABLED
+
+	//revive:disable-next-line:var-naming ALL_CAPS
+	ERROR_NOT_ALL_ASSIGNED syscall.Errno = windows.ERROR_NOT_ALL_ASSIGNED
```
Comment (MichaelEischer) [inline internal/fs/sd_windows.go:234]: The upstream version uses `windows.Errno`.

File: internal/fs/sd_windows.go [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+	}
+
+	if lowerPrivileges {
+		err = setNamedSecurityInfoLow(filePath, dacl)
+	} else {
+		err = setNamedSecurityInfoHigh(filePath, owner, group, dacl, sacl)
+	}
+
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
```
Comment (MichaelEischer) [inline internal/fs/sd_windows.go:114]: Should also check for `!lowerPrivileges`

File: internal/fs/sd_windows.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go:56]: I agree that it probably won't break anything. However, the golang race detector will probably complain about some of that code. The simplest solution would be to use an atomic boolean instead: https://pkg.go.dev/sync/atomic#Bool .

File: internal/fs/sd_windows.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: > I guess SDs won't work when moving between 32 and 64 bit systems.

According to https://nostarch.com/download/WindowsSecurityInternals_Chapter5.pdf the self-relative SDs, which are used here, are encoded the same on 32/64 bit systems. That also matches that https://learn.microsoft.com/en-us/windows/win32/secauthz/absolute-and-self-relative-security-descriptors says that the format can be stored/transmitted.

So I'm not entirely happy with how we have to serialize the SDs. But as the code is based on https://github.com/microsoft/go-winio/blob/3c9576c9346a1892dee136329e7e15309e82fb4f/sd.go#L118 , the approach is probably fine.

File: internal/fs/sd_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:352]: That's a little harsh. I wrote this code 2 years ago before I decided to submit it back to restic.

The upstream code was changed only last month - https://github.com/microsoft/go-winio/commit/3c9576c9346a1892dee136329e7e15309e82fb4f
Eg.
	- r0, _, _ := syscall.Syscall(procHcsFormatWritableLayerVhd.Addr(), 1, uintptr(handle), 0, 0)
	+ r0, _, _ := syscall.SyscallN(procHcsFormatWritableLayerVhd.Addr(), uintptr(handle))
	
The changes I made 2 years ago were to fix the deprecations - 
Deprecated: Use SyscallN instead.	

Extra 0s at the end of SyscallN are ignored and hence this slip wasn't caught, and it also doesn't cause any bugs, though I agree it should be cleaned.

Also, I get your point on including links to the source and will try to do that, this was missed because the code was created from upstream a long time ago.

File: internal/fs/sd_windows.go [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+	var _p0 *uint16
+	_p0, err = syscall.UTF16PtrFromString(systemName)
+	if err != nil {
+		return
+	}
+	return _lookupPrivilegeDisplayName(_p0, name, buffer, size, languageID)
+}
+
+func _lookupPrivilegeDisplayName(systemName *uint16, name *uint16, buffer *uint16, size *uint32, languageID *uint32) (err error) {
+	r1, _, e1 := syscall.SyscallN(procLookupPrivilegeDisplayNameW.Addr(), uintptr(unsafe.Pointer(systemName)), uintptr(unsafe.Pointer(name)), uintptr(unsafe.Pointer(buffer)), uintptr(unsafe.Pointer(size)), uintptr(unsafe.Pointer(languageID)), 0)
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:375]: Fixed

File: internal/fs/sd_windows.go [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+	var _p0 *uint16
+	_p0, err = syscall.UTF16PtrFromString(systemName)
+	if err != nil {
+		return
+	}
+	return _lookupPrivilegeName(_p0, luid, buffer, size)
+}
+
+func _lookupPrivilegeName(systemName *uint16, luid *uint64, buffer *uint16, size *uint32) (err error) {
+	r1, _, e1 := syscall.SyscallN(procLookupPrivilegeNameW.Addr(), uintptr(unsafe.Pointer(systemName)), uintptr(unsafe.Pointer(luid)), uintptr(unsafe.Pointer(buffer)), uintptr(unsafe.Pointer(size)), 0, 0)
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:392]: Fixed

File: internal/fs/sd_windows.go (hunk 20 above)
Comment (aneesh-n) [reply internal/fs/sd_windows.go:234]: This deprecation clean-up was done upstream 9 months ago after this feature code was written. 
Have updated the code now

File: internal/fs/sd_windows.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:291]: DisableBackupPrivileges and DisableRestorePrivileges were removed. This one got left out.
Removed now.

File: internal/fs/sd_windows.go [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+}
+
+func adjustTokenPrivileges(token windows.Token, releaseAll bool, input *byte, outputSize uint32, output *byte, requiredSize *uint32) (success bool, err error) {
+	var _p0 uint32
+	if releaseAll {
+		_p0 = 1
+	}
+	r0, _, e1 := syscall.SyscallN(procAdjustTokenPrivileges.Addr(), uintptr(token), uintptr(_p0), uintptr(unsafe.Pointer(input)), uintptr(outputSize), uintptr(unsafe.Pointer(output)), uintptr(unsafe.Pointer(requiredSize)))
+	success = r0 != 0
+	if !success {
```
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go:359]: This windows api always returns an error even in case of success, warnings (partial success) and error cases.

Full success - When we call this with admin permissions, it returns DNS_ERROR_RCODE_NO_ERROR (0). This gets translated to errErrorEinval and ultimately in adjustTokenPrivileges, it gets ignored.

Partial success - If we call this api without admin privileges, privileges related to SACLs do not get set and though the api returns success, it returns an error - golang.org/x/sys/windows.ERROR_NOT_ALL_ASSIGNED (1300)
This gets propagated up the call hierarchy and then gets printed as a debug log in enableRestorePrivilege or enableBackupPrivilege. 
Now that we have changed the handling in those functions to print debug log, it will start printing debug logs even in case ERROR_NOT_ALL_ASSIGNED was thrown, which is a warning and in our case is expected in case we run restic without admin permissions. In that case adjustPrivileges will return a PrivilegeError even though other privileges other than SACL access were still set.

Instead of eating the warning on one extreme or throwing a Privilege error on the other extreme, I can handle this warning in adjustPrivileges and add a debug statement - 

debug.Log("Not all requested privileges were fully set: %v. AdjustTokenPrivileges returned warning: %v", privileges, err)
This debug statement will get logged when we call restic without admin permissions.

File: internal/fs/sd_windows.go (hunk 21 above)
Comment (aneesh-n) [reply internal/fs/sd_windows.go:114]: Fixed

File: internal/fs/sd_windows.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (aneesh-n) [reply internal/fs/sd_windows.go]: Agree, it makes sense to add and option to include/exclude SDs during restore. This can be added in this PR or added as a new PR.

[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: let's keep that option out of this PR. It's already large enough as it is ;-)

File: internal/fs/sd_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go:358]: > That's a little harsh.

Let's just say that I was rather annoyed after noticing that I have to diff the syscall wrappers and search for the corresponding source code.

> I wrote this code 2 years ago before I decided to submit it back to restic.

That perfectly explains why the code looks the way it does. Thanks for contributing those changes!

File: internal/fs/sd_windows.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply internal/fs/sd_windows.go]: The debug.Log is fine.

Judging from https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-adjusttokenprivileges `if true {` is in fact the correct variant (even though it looks weird. However, calling errnoErr doesn't really make sense in that case), as `GetLastError` can return either ERROR_SUCCESS or ERROR_NOT_ALL_ASSIGNED if adjustTokenPrivileges succeeded.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

**Comment by zumm** [comment]
Is there a way to disable this feature? I'm having tough times synchronizing files between machines.
